mlflow/types/schema.py

```diff
@@ -116,84 +116,189 @@ def __repr__(self) -> str:
             return "{name}: {type}".format(name=repr(self.name), type=repr(self.type))
 
 
+class TensorSpec(object):
+    """
+    Specification used to represent a dataset stored as a Tensor.
+    """
+
+    def __init__(
+        self,
+        type: np.dtype,  # pylint: disable=redefined-builtin
+        shape: tuple,
+        name: Optional[str] = None,
+    ):
+        self._name = name
+        if not isinstance(type, np.dtype):
+            raise TypeError(
+                "Expected `type` to be instance of `{0}`, received `{1}`".format(
+                    np.dtype, type.__class__
+                )
+            )
+        if not isinstance(shape, tuple):
+            raise TypeError(
+                "Expected `shape` to be instance of `{0}`, received `{1}`".format(
+                    tuple, shape.__class__
+                )
+            )
+        self._type = type
+        self._shape = shape
+
+    @property
+    def type(self) -> np.dtype:
+        """
+        A unique character code for each of the 21 different numpy built-in types.
+        See https://numpy.org/devdocs/reference/generated/numpy.dtype.html#numpy.dtype for details.
+        """
+        return self._type
+
+    @property
+    def name(self) -> Optional[str]:
+        """The tensor name or None if the tensor is unnamed."""
+        return self._name
+
+    @property
+    def shape(self) -> tuple:
+        """The tensor shape"""
+        return self._shape
+
+    def to_dict(self) -> Dict[str, Any]:
+        if self.name is None:
+            return {"type": self.type.name, "shape": self.shape}
+        else:
+            return {"name": self.name, "type": self.type.name, "shape": self.shape}
+
+    @classmethod
+    def from_json_dict(cls, **kwargs):
+        """
+        Deserialize from a json loaded dictionary.
+        The dictionary is expected to contain `type` and `shape` keys.
+        """
+        tensor_type = np.dtype(kwargs["type"])
+        tensor_shape = tuple(kwargs["shape"])
+        return cls(tensor_type, tensor_shape, kwargs["name"] if "name" in kwargs else None)
+
+    def __eq__(self, other) -> bool:
```

Comment:
we should check if other is TensorSpec. just in case someone compares schema with colspec with schema with tensorspec